Python/import.c

```diff
@@ -1648,13 +1648,15 @@ resolve_name(PyThreadState *tstate, PyObject *name, PyObject *globals, int level
                 goto error;
             }
 
```

Comment:
Instead of the implicit failure by faking that the dot was found at 0, add a clause of the failure case and then make the latter guard just be an else:

```suggestion
            else if (dot == -1) {
                _PyErr_SetString(tstate, PyExc_ImportError,
                         "attempted relative import "
                         "with no known parent package");
                goto error;
            }
            else {
```
This more clearly communicates what you're trying to accomplish. Basically there's no need to be clever here as it isn't a performance win in a critical code path.

Reply:
I've pushed another commit to tidy this up.